### ISSUE TITLE
Allow to configure directories for Test Explorer

### ISSUE THREAD

**Issue by slacaze** [golang/vscode-go#2383]
**Is your feature request related to a problem? Please describe.**

We have an automated CI pipeline where all build files should be contained in the project workspace.

GOPATH is configured to be "./build", with a directory tree like:

```
project_root/
├── go.mod
├── cmd/
├── pkg/
└── target/
    ├─ bin/
    └─ pkg/
```

This is causing Test Explorer to show not only all the test of the project, but all the tests of the project transitive dependencies. It also impact performance (time to render test list).

**Describe the solution you'd like**

The current behavior of Test Explorer is akin to `go test ./... -list` (I don't actually know how Test Explorer is populated).

We would greatly benefit from being able to profite custom package lists, like `go test ./cmd/... ./pkg/... -list`.

Alternatively, a folder exclusion list, or a list of directory to consider would also be fine.

**Describe alternatives you've considered**

We might be able to use regexp once the Test Explorer is populated, but the performance would still be an issue, and that is not a setting we could save in the repo, to be shared by all users.

We thought about moving everything down one level:

```
project_root/
├── src/
│   ├── go.mod
│   ├── cmd/
│   └── pkg/
└── target/
    ├── bin/
    └── pkg/
```

But the repo then looks odd, as it just has a src/ folder (the build folder is only created during the CI run). Also imports are now "project_path/src/pkg/foo", which also looks odd.

We thought about moving the GOPATH outside the project root, but that would not be inline with our other build pipelines, and would go against keeping each build area isolated.

**Additional context**

This is obviously very inspired by Maven as all of our build tools are based on Maven plugins.

**Comment by hyangah** [comment]
Related golang/vscode-go#1739 and @firelizzard18 's pending cl https://go-review.googlesource.com/c/vscode-go/+/351172

In addition to that we can also exclude GOMODCACHE from scanning paths completely.

cc @firelizzard18 

@slacaze Can you elaborate on this?

> We thought about moving the GOPATH outside the project root, but that would not be inline with our other build pipelines, and would go against keeping each build area isolated.

My personal recommendation is to stop placing `GOPATH` inside the project workspace. With Go Modules, `GOPATH` is no longer served as the location for dependency source code (`GOMODCACHE` is where the transitive dependencies' source code may be installed and happened to be default to `GOPATH/pkg/mod` but this can be configured to be in a different directory).

**Comment by slacaze** [comment]
@hyangah Thank you for taking the time to look into this.

I can confirm the other issue you've mentioned would address our needs. Excluding GOMODCACHE (I would need to read more about this, I didn't know about it) sounds like it would also address our issues.

Regarding your question, and making no claim that our approach is good or bad, we're following Maven style builds (and I suppose node/npm, maybe others) to (as much as possible) contain the changes to the system made by a build restrained to the build folder. In Maven, everything is in `target`, you delete it, you remove all traces of the build. In node/npm, `node install` creates a `node_modules` (if I remember correctly) at the root of the project. We thought to follow the same pattern for our golang builds.

We've made no research as to whether golang GOPATH or GOMODCACHE would be more resilient to concurrent build from different agents on a single system (in contrast, running parallel Maven builds sharing a `target` directory, or node projects sharing the same `node_modules` would be likely disastrous).

We simply believed that as long as we did the same with golang, we would stay out of trouble, and retain consistencies across our build targets.

I hope this makes sense.
